src/util/fs.cc

```diff
@@ -121,4 +127,377 @@ bool os_portable_copy_file_to_dir(const std::string& src,
+    ) {
+      _ret.failure = execute_result::EXEC;
+      return _ret;
+    }
+  CloseHandle(pi.hProcess);
+  CloseHandle(pi.hThread);
+  _ret.failure = execute_result::NONE;
+  return _ret; 
+#else
+  // set up the pipe to transfer subprocess's information before exec
+  // so close on exec
+  gettimeofday(&Time1, NULL);
+
+  pipe2(pipefd, O_CLOEXEC); // 
+  // on *nix, spawn a child process to do this
+  pid_t pid = fork();
+  
+  if (pid == -1) {
+    _ret.failure = execute_result::FORK;
+    // not able to create no process
+    return _ret;
+  }
+
+  // now forked ...
+  if (pid == 0) {
+
+    if(timeout != 0) // only if we want to have the time-out feature
+      setpgid(0,0); // creates a new proces group 
+    
+    close(pipefd[0]); // close the read end
+    int report_to_parent;
+
+    // The child
+    // will replace the image and execute the bash
+    ILA_INFO<<"Execute subprocess: [" << cmdline << "]";
+    if (! redirect_output_file.empty()) {
+      ILA_INFO<<"Redirect to:" << redirect_output_file;
+      int fd;
+      fd = open(redirect_output_file.c_str(), O_WRONLY | O_CREAT | O_TRUNC , S_IRUSR | S_IWUSR );
+      if (fd < 0) {
+        perror(NULL);
+        ILA_ERROR << "Failed to open " << redirect_output_file;
+        report_to_parent = execute_result::PREIO;
+        write(pipefd[1], (void *) & report_to_parent, sizeof(report_to_parent));
+        exit(1);
+      }
+      if (rdt & redirect_t::STDOUT)
+        dup2(fd, STDOUT_FILENO);
+      if (rdt & redirect_t::STDERR)
+        dup2(fd, STDERR_FILENO);
+      close(fd);
+    }
+    
+    // this is memory leak, but I have no other way...
+    // hope this will be reclaimed by OS
+    // + 1 for NULL
+
+    const int MAX_ARG = 64;
+    char * argv[MAX_ARG];
+    if (cmdargs.size() >= MAX_ARG) {
+      ILA_ERROR <<"Too many args";
+      report_to_parent = execute_result::ARG;
+      write(pipefd[1], (void *) & report_to_parent, sizeof(report_to_parent));
+      exit(1);
+    }
+    //ILA_ASSERT(cmdargs.size() < MAX_ARG) << "Too many args";
+
+    for(auto it = cmdargs.begin(); it != cmdargs.end(); ++ it) {
+      // this is memory leak, I know, but what can I do ?
+      argv[it-cmdargs.begin()] = new char[it->size() + 1];
+      strcpy(argv[it-cmdargs.begin()], it->c_str());
+    }
+    argv[ cmdargs.size() ] = NULL;
+
+    report_to_parent = execute_result::NONE;
+    write(pipefd[1], (void *) & report_to_parent, sizeof(report_to_parent));
+
+    int ret = execvp(cmdargs[0].c_str(), argv);
+    // if not successful
+    if (ret == -1) {
+      report_to_parent = execute_result::NONE;
+      write(pipefd[1], (void *) & report_to_parent, sizeof(report_to_parent));
+    }
+    exit(ret);
+  } else {
+    // The parent will wait for its end
+    int infop;
+    int child_report;
+    struct sigaction new_act;
+    struct sigaction old_act;
+
+    close(pipefd[1]); // close the write end
+    read(pipefd[0], (void *) &child_report, sizeof(child_report));
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Check buffer boundaries if used in a loop including recursive loops (CWE-120, CWE-20).  ](https://app.codacy.com/app/Bo-Yuan-Huang/ILAng/pullRequest?prid=4555809)